File: 1016-subarray-sums-divisible-by-k/1016-subarray-sums-divisible-by-k.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        int n = nums.size();
        int cnt = 0;
        unordered_map<int,int> mp;
        mp[0] = 1;
        int sm = 0;
        for(int i = 0; i<n; i++){
            sm += nums[i];
            int rem = sm % k;

            if(rem < 0){
                rem += k;
            }
            if(mp.find(rem) != mp.end()){
                cnt += mp[rem];
            }
            mp[rem]++;
        }
        return cnt;
    }
};

File: 1016-subarray-sums-divisible-by-k/1016-subarray-sums-divisible-by-k.java
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        int n = nums.length;
        Map<Integer,Integer> mp = new HashMap<>();
        int cntSubArray = 0;

        mp.put(0,1);
        int sm = 0, rem = 0;
        for(int i = 0; i<n; i++){
            sm += nums[i];
            rem = sm%k;
            if(rem < 0){
                rem += k;
            }
            if(mp.containsKey(rem)){
                cntSubArray += mp.get(rem);
            }
            mp.put(rem, mp.getOrDefault(rem,0)+1);
        }
        return cntSubArray;
    }
}